src/mscorlib/src/System/IO/UnmanagedMemoryAccessor.cs

```diff
@@ -789,25 +738,55 @@ public class UnmanagedMemoryAccessor : IDisposable {
             }
         }
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private unsafe void UnsafeWriteInt32(byte* pointer, Int32 value)
+        {
+            // check if pointer is aligned
+            if (((int)pointer & (sizeof(Int32) - 1)) == 0)
+            {
+                *((Int32*)pointer) = value;
+            }
+            else
+            {
+                *(pointer) = (byte)value;
+                *(pointer + 1) = (byte)(value >> 8);
+                *(pointer + 2) = (byte)(value >> 16);
+                *(pointer + 3) = (byte)(value >> 24);
+            }
+        }
+
         [System.Security.SecuritySafeCritical]  // auto-generated
         public void Write(Int64 position, Decimal value) {
             int sizeOfType = sizeof(Decimal);
             EnsureSafeToWrite(position, sizeOfType);
 
-            byte[] decimalArray = new byte[16];
-            Decimal.GetBytes(value, decimalArray);
+            unsafe
+            {
+                byte* pointer = null;
+                try
+                {
+                    _buffer.AcquirePointer(ref pointer);
```

Comment:
There should be `PrepareConstrainedRegions` in frong of every `AcquirePointer` to make correct for desktop CERs.